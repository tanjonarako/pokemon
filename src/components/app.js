import React, { Component } from 'react'
import List from './list'
import Favourite from './favourite'
import '../stylesheet/app.css'

class App extends Component {
  render () {
    return (
      <div className='app'>
        <Favourite />
        <List />
      </div>
    )
  }
}

export default App
