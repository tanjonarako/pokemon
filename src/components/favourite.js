import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const mapDispatchToProps = (dispatch) => ({
  getFav: () => dispatch({
    type: "getFav"
  })
});

export class Favourite extends Component {
  static propTypes = {
    list: PropTypes.array,
    getFav: PropTypes.func
  }

  componentDidMount() {
    this.props.getFav()
  }

  render() {
    const favourites = this.props.list
    return (
      <div>
        {favourites && favourites.length > 0 && <div className="favourite px-2 py-2">
          <h1>Favourite</h1>
          <ul>
            {favourites.map((favouriteId) => {
              return <li key={favouriteId}>{favouriteId}</li>
            })}
          </ul>
          <div className='separator'></div>
        </div>}
      </div>
    )
  }
}

Favourite.defaultProps = { list: [] }

export default connect(
  state => ({
    list: state.favourites,
  }),
  mapDispatchToProps
)(Favourite)
