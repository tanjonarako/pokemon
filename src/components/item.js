import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const fetchItemAction = (name) => ({
  type: "updateListItem",
  payload: {
    request: {
      url: 'https://pokeapi.co/api/v2/pokemon/' + name
    }
  }
});

const mapDispatchToProps = (dispatch) => ({
  doFetch: name => dispatch(fetchItemAction(name)),
  fav: (id) => dispatch({
    type: "addToFav",
    data: id
  })
});

export class Item extends Component {
  static propTypes = {
    doFetch: PropTypes.func
  }

  state = {
    opened: false
  }

  toggle = () => {
    const { opened } = this.state
    this.setState({
      opened: !opened
    })
  }

  render() {
    const { name, weight, sprites, abilities} = this.props.item
    const { opened } = this.state
    const itemClick = name => {
      this.toggle()
      this.props.doFetch(name)
    }
    return (
      <div className='item-details col-xs-12 col-sm-3 col-md-3 col-lg-3 px-0 py-2'>
        <a href='#' onClick={() => itemClick(name)} className="item-link">
          <h2 className='name'>{name}</h2>
          {opened && <div className='more-info mb-4'>
            {!!sprites && <img src={sprites.front_default}  className="img-thumbnail"/>}
            <div className='separator'></div>
            {!!weight && <p>Weight: {weight}</p>}
            {!!abilities && <ul>Abilities: {abilities.map((item) => {
              return (
                <li key={item.ability.name}>{item.ability.name}</li>
              )
            })}</ul>}
          </div>}
        </a>
        <div>
          <button className="btn btn-primary" onClick={() => this.props.fav(name)}>Add to my collection</button>
        </div>
      </div>
    );
  }
}

Item.defaultProps = { item: {} }

export default connect(
  (state, props) => ({
    item: state.pokemonList.byIds[props.id]
  }),
  mapDispatchToProps
)(Item)
