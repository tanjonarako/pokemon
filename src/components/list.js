import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Item from './item';

const fetchListAction = () => ({
  type: "updateList",
  payload: {
    request: {
      url: 'https://pokeapi.co/api/v2/pokemon'
    }
  }
});

const mapDispatchToProps = (dispatch) => ({
  doFetch: () => dispatch(fetchListAction()),
  fav: (id) => dispatch({
    type: "addToFav",
    data: id
  })
});

export class List extends Component {
  static propTypes = {
    list: PropTypes.array,
    doFetch: PropTypes.func
  }

  componentDidMount() {
    this.props.doFetch()
  }

  render() {
    const list = this.props.list
    return (
      <div className="list">
        {list.map((id) => (
          <Item key={id} id={id}/>
        ))}
      </div>
    );
  }
}

List.defaultProps = { list: [] }

export default connect(
  state => ({
    list: state.pokemonList.ids
  }),
  mapDispatchToProps
)(List)
