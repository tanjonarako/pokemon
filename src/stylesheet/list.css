.list {
  margin:0 auto;  
  display: flex;
  flex-wrap: wrap;
  padding: 10px;
  justify-content: space-between;
}
