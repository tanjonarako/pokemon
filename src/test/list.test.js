import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { List } from '../components/list'

configure({ adapter: new Adapter() })

describe('List', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<List />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a list', () => {
    const list = ['abc', 'def']
    const wrapper = shallow(<List list={list} />)
    expect(wrapper).toMatchSnapshot()
  })
})
