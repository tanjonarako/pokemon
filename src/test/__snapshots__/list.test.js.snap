// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render a list 1`] = `
<div>
  <div
    className="list"
  >
    <Connect(Item)
      id="abc"
      key="abc"
    />
    <Connect(Item)
      id="def"
      key="def"
    />
  </div>
  <nav>
    <ul
      className="pagination"
    >
      <li
        className="page-item px-0 disabled"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          Previous
        </a>
      </li>
      <li
        className="page-item disabled"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          Next
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`List should render correctly 1`] = `
<div>
  <div
    className="list"
  />
  <nav>
    <ul
      className="pagination"
    >
      <li
        className="page-item px-0 disabled"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          Previous
        </a>
      </li>
      <li
        className="page-item disabled"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          Next
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
